### PR TITLE
James 1784 - Attachment : handle query string on downloadUrl to manage signature (work on filters) 

### PR DESCRIPTION
**Pull request by aduprat**
Rebase + take care of comment published in #308 
Commit order is wrong on GitHub, real order is:
JAMES-1784 Authentication strategy should be responsible of extracting the authentication part from the request
JAMES-1784 Remove checkAuthorizationHeader in AuthenticationStrategy
JAMES-1784 Add QueryParameterAccessTokenAuthenticationStrategy
JAMES-1784 Add integration tests when downloading attachment with access token

### CODE REVIEW

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/AuthenticationFilterTest.java [hunk 1]
```diff
@@ -129,12 +147,67 @@ private FakeAuthenticationStrategy(boolean isAuthorized) {
...
         @Override
         public MailboxSession createMailboxSession(HttpServletRequest httpRequest) {
+            if (!isAuthorized) {
+                throw new MailboxSessionCreationException(null);
+            }
+            return new FakeMailboxSession();
+        }
+    }
+
+    private static class FakeMailboxSession implements MailboxSession {
```
Comment (rouazana) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/AuthenticationFilterTest.java]: Any added value on org.apache.james.mailbox.mock.MockMailboxSession?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java [hunk 2]
```diff
@@ -49,25 +53,36 @@
...
-        this.blobIdByAttachmentId = new HashMap<>();
+        this.blobIdByAttachmentId = Maps.newHashMap();
+        this.attachmentAccessTokens = Maps.newHashMap();
     }
 
     @Given("^\"([^\"]*)\" mailbox \"([^\"]*)\" contains a message \"([^\"]*)\" with an attachment \"([^\"]*)\"$")
     public void appendMessageWithAttachmentToMailbox(String user, String mailbox, String messageId, String attachmentId) throws Throwable {
         MailboxPath mailboxPath = new MailboxPath(MailboxConstants.USER_NAMESPACE, user, mailbox);
 
+        // blodId = "4000c5145f633410b80be368c44e1c394bff9437"
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: you can remove this comment?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java [hunk 3]
```diff
@@ -49,25 +53,36 @@
...
     private ValidatableResponse validatableResponse;
+    private Map<AttachmentAccessTokenKey, AttachmentAccessToken> attachmentAccessTokens;
 
     @Inject
     private DownloadStepdefs(MainStepdefs mainStepdefs, UserStepdefs userStepdefs) {
         this.mainStepdefs = mainStepdefs;
         this.userStepdefs = userStepdefs;
         this.attachmentsByMessageId = ArrayListMultimap.create();
-        this.blobIdByAttachmentId = new HashMap<>();
+        this.blobIdByAttachmentId = Maps.newHashMap();
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: why don't you like the `new HashMap<>()` ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java [hunk 4]
```diff
@@ -95,38 +110,147 @@ public void optionDownload(String username) throws Throwable {
...
     }
 
     @When("^\"([^\"]*)\" downloads \"([^\"]*)\"$")
     public void downloads(String username, String attachmentId) throws Throwable {
         String blobId = blobIdByAttachmentId.get(attachmentId);
+        response = requestWithAuthentication(blobId, username)
+                .get("/download/" + blobId);
+    }
+
+    private RequestSpecification requestWithAuthentication(String blobId, String username) {
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: the method name is more generic than its actual use : it's specific to download API

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java [hunk 5]
```diff
@@ -95,38 +110,147 @@ public void optionDownload(String username) throws Throwable {
...
         if (accessToken != null) {
-            with.header("Authorization", accessToken.serialize());
+            return with().header("Authorization", accessToken.serialize());
+        }
+        return with();
+    }
+
+    @When("^\"([^\"]*)\" is trusted for attachment \"([^\"]*)\"$")
+    public void attachmentAccessTokenFor(String username, String attachmentId) throws Throwable {
+        userStepdefs.connectUser(username);
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: how making `username` trusted for an attachment access is related to connecting him ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java [hunk 6]
```diff
@@ -95,38 +110,147 @@ public void optionDownload(String username) throws Throwable {
...
+        trustForBlobId(blobIdByAttachmentId.get(attachmentId), userStepdefs.tokenByUser.get(username));
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: trust should take the username, not the token

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java [hunk 7]
```diff
@@ -95,38 +110,147 @@ public void optionDownload(String username) throws Throwable {
...
+    private void trustForBlobId(String blobId, AccessToken accessToken) {
+        String serializedAttachmentAccessToken = 
+            with()
+                .header("Authorization", accessToken.serialize())
+            .post("/download/" + blobId)
+            .then()
+                .extract()
+                .asString();
+        attachmentAccessTokens.put(
+                new AttachmentAccessTokenKey(userStepdefs.lastConnectedUser, blobId),
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: this line is broken : what if it's not the lastConnectedUser ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java [hunk 8]
```diff
@@ -95,38 +110,147 @@ public void optionDownload(String username) throws Throwable {
...
+                    serializedAttachmentAccessToken,
+                    blobId));
+    }
+
+    @When("^\"([^\"]*)\" downloads \"([^\"]*)\" with a valid authentication token$")
+    public void downloadsWithValidToken(String username, String attachmentId) throws Throwable {
+        String blobId = blobIdByAttachmentId.get(attachmentId);
         AccessToken accessToken = userStepdefs.tokenByUser.get(username);
+
+        response = addQueryParameterAuthentication(username, blobId, accessToken)
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: add on what ?

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/AuthenticationFilterTest.java (hunk 1 above)
Comment (aduprat) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/AuthenticationFilterTest.java]: MockMailboxSession has some behaviour, i need none.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java (hunk 2 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: OK

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java (hunk 3 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: Old reflex.

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/AuthenticationFilterTest.java (hunk 1 above)
Comment (rouazana) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/AuthenticationFilterTest.java]: then why not mock(MailboxSession.class)?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java (hunk 4 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: This method returns an authenticated request ready to use.
It's handling both header & query parameter authentication mechanisms, as we are testing both.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/AuthenticationFilter.java [hunk 9]
```diff
@@ -71,9 +68,10 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
 
         try {
             HttpServletRequest requestWithSession = authMethods.stream()
-                    .filter(auth -> auth.checkAuthorizationHeader(getAuthHeaders(httpRequest)))
+                    .map(auth -> createSession(auth, httpRequest))
+                    .filter(Optional::isPresent)
                     .findFirst()
-                    .map(auth -> addSessionToRequest(httpRequest, createSession(auth, getAuthHeaders(httpRequest))))
+                    .map(mailboxSession -> addSessionToRequest(httpRequest, mailboxSession))
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/AuthenticationFilter.java]: if you put  `map`  before `findFirst` you would not get a Optional<MailboxSession> as addSessionToRequest argument

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java (hunk 5 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: In order to be trusted, we have to connect first, then access to the POST method.
Then the sentence:
`is trusted for attachment`
is more logic when testing the GET method than:
`is connected`
`is trusted for attachment`

As the user doesn't have to be `connected`.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/AuthenticationFilter.java [hunk 10]
```diff
@@ -84,19 +82,17 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
...
-    private HttpServletRequest addSessionToRequest(HttpServletRequest httpRequest, MailboxSession mailboxSession) {
-        httpRequest.setAttribute(MAILBOX_SESSION, mailboxSession);
+    private HttpServletRequest addSessionToRequest(HttpServletRequest httpRequest, Optional<MailboxSession> mailboxSession) {
+        httpRequest.setAttribute(MAILBOX_SESSION, mailboxSession.get());
         return httpRequest;
     }
 
-    private MailboxSession createSession(AuthenticationStrategy authenticationMethod, Stream<String> authorizationHeaders) {
-        return authenticationMethod.createMailboxSession(authorizationHeaders);
+    private Optional<MailboxSession> createSession(AuthenticationStrategy authenticationMethod, HttpServletRequest httpRequest) {
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/AuthenticationFilter.java]: It's a rare case where I would use Stream to be able to flatmap afterward without Optional Hell

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java (hunk 6 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: OK

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java (hunk 8 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: OK

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java (hunk 7 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: Oops

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/AuthenticationFilterTest.java (hunk 1 above)
Comment (aduprat) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/AuthenticationFilterTest.java]: OK

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/AuthenticationFilter.java (hunk 9 above)
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/AuthenticationFilter.java]: OK

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/AuthenticationFilter.java (hunk 10 above)
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/AuthenticationFilter.java]: OK

**Comment by rouazana** [comment]
:+1:

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java (hunk 4 above)
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: But query parameter access token doesn't make sense for anything else than download

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java (hunk 5 above)
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: Ok, so : 

> In order to be trusted, we have to connect first

and 

> As the user doesn't have to be connected

Then maybe you should not reuse `connectUser`

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java (hunk 4 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: OK

**Comment by mbaechler** [comment]
:+1:

**Comment by rouazana** [comment]
:+1:

**Comment by aduprat** [comment]
test this please

**Pull request merged**